presto-main/src/main/java/com/facebook/presto/operator/FilterAndProjectOperator.java

```diff
@@ -96,24 +94,29 @@ public final void addInput(Page page)
     public final Page getOutput()
     {
         if (!pageBuilder.isFull() && currentPage != null) {
-            if (columnarProcessingDictionaryEnabled) {
-                Page page = processor.processColumnarDictionary(operatorContext.getSession().toConnectorSession(), currentPage, getTypes());
-                currentPage = null;
-                currentPosition = 0;
-                return page;
-            }
-            else if (columnarProcessingEnabled) {
-                Page page = processor.processColumnar(operatorContext.getSession().toConnectorSession(), currentPage, getTypes());
-                currentPage = null;
-                currentPosition = 0;
-                return page;
-            }
-            else {
-                currentPosition = processor.process(operatorContext.getSession().toConnectorSession(), currentPage, currentPosition, currentPage.getPositionCount(), pageBuilder);
-                if (currentPosition == currentPage.getPositionCount()) {
+            switch (processingOptimization) {
+                case FeaturesConfig.ProcessingOptimization.COLUMNAR: {
+                    Page page = processor.processColumnar(operatorContext.getSession().toConnectorSession(), currentPage, getTypes());
+                    currentPage = null;
+                    currentPosition = 0;
+                    return page;
+                }
+                case FeaturesConfig.ProcessingOptimization.COLUMNAR_DICTIONARY: {
+                    Page page = processor.processColumnarDictionary(operatorContext.getSession().toConnectorSession(), currentPage, getTypes());
                     currentPage = null;
                     currentPosition = 0;
+                    return page;
+                }
+                case FeaturesConfig.ProcessingOptimization.DISABLED: {
+                    currentPosition = processor.process(operatorContext.getSession().toConnectorSession(), currentPage, currentPosition, currentPage.getPositionCount(), pageBuilder);
+                    if (currentPosition == currentPage.getPositionCount()) {
+                        currentPage = null;
+                        currentPosition = 0;
+                    }
+                    break;
                 }
+                default:
+                    throw new IllegalStateException();
```

Comment:
`throw new IllegalStateException(String.format("Found unexpected value %s for processingOptimization", processingOptimization))`